locator/token_metadata.cc

```diff
@@ -783,6 +790,44 @@ size_t token_metadata_impl::count_normal_token_owners() const {
     return _normal_token_owners.size();
 }
 
+void token_metadata_impl::for_each_token_owner(std::function<void(const node*)> func) const {
+    auto nodes = _topology.get_nodes();
+    for (const auto* np : nodes) {
+        if (is_normal_token_owner(np->host_id())) {
+            func(np);
+        }
+    }
+}
+
+std::unordered_map<sstring, std::unordered_set<inet_address>> token_metadata_impl::get_datacenter_token_owners() const {
+    std::unordered_map<sstring, std::unordered_set<inet_address>> datacenter_token_owners;
+    for (const auto& [dc, ips] : _topology.get_datacenter_endpoints()) {
+        std::unordered_set<inet_address> token_owners;
+        std::copy_if(ips.begin(), ips.end(), std::inserter(token_owners, token_owners.begin()), [this] (const auto& ip) {
+            return is_normal_token_owner(get_host_id_if_known(ip).value_or(host_id{}));
```

Comment:
calling `get_host_id_if_known` is wasteful.
locator::topology knows the host_id of endpoints, so you can use `for_each_node` instead to get to the node host_id() and dc_rack().dc

Reply:
I used `for_each_node` as you suggested, and not only did I get rid of `get_host_id_if_known`, but the code is also less verbose.